clcache.py

```diff
@@ -1600,57 +1574,89 @@ def processCompileRequest(cache, compiler, args):
-                Statistics.registerSourceChangedMiss)
+        if manifest:
+            for entryIndex, entry in enumerate(manifest.entries()):
+                # NOTE: command line options already included in hash for manifest name
+                try:
+                    includesContentHash = ManifestRepository.getIncludesContentHashForFiles(
+                        [expandBasedirPlaceholder(path) for path in entry.includeFiles])
+
+                    if entry.includesContentHash == includesContentHash:
+                        cachekey = entry.objectHash
+                        assert cachekey is not None
+                        # Move manifest entry to the top of the entries in the manifest
+                        manifest.touchEntry(entryIndex)
+                        manifestSection.setManifest(manifestHash, manifest)
+
+                        artifactSection = cache.compilerArtifactsRepository.section(cachekey)
+                        with artifactSection.lock:
+                            if artifactSection.hasEntry(cachekey):
+                                return processCacheHit(cache, objectFile, cachekey)
+
+                except IncludeNotFoundException:
+                    pass
 
-        for entryIndex, entry in enumerate(manifest.entries()):
-            # NOTE: command line options already included in hash for manifest name
-            try:
-                includesContentHash = ManifestRepository.getIncludesContentHashForFiles(
-                    [expandBasedirPlaceholder(path) for path in entry.includeFiles])
+            unusableManifestMissReason = Statistics.registerHeaderChangedMiss
+        else:
+            unusableManifestMissReason = Statistics.registerSourceChangedMiss
+
+    if artifactSection is None:
+        stripIncludes = False
+        if '/showIncludes' not in cmdLine:
+            cmdLine = list(cmdLine)
+            cmdLine.insert(0, '/showIncludes')
+            stripIncludes = True
+    compilerResult = invokeRealCompiler(compiler, cmdLine, captureOutput=True)
+    if artifactSection is None:
+        includePaths, compilerOutput = parseIncludesSet(compilerResult[1], sourceFile, stripIncludes)
+        compilerResult = (compilerResult[0], compilerOutput, compilerResult[2])
 
-                if entry.includesContentHash == includesContentHash:
-                    cachekey = entry.objectHash
-                    assert cachekey is not None
-                    # Move manifest entry to the top of the entries in the manifest
-                    manifest.touchEntry(entryIndex)
-                    manifestSection.setManifest(manifestHash, manifest)
+    with manifestSection.lock:
+        if artifactSection is not None:
+            return ensureArtifactsExist(cache, artifactSection, cachekey, unusableManifestMissReason,
+                                        objectFile, compilerResult)
+
+        entry = createManifestEntry(manifestHash, includePaths)
```

Comment:
This line computes the hashes for all the includes and may be slow too. It would be nice to move it outside of the critical section too.